### PR TITLE
Open any API URL with your browser and a nice UI shows up

### PR DESCRIPTION
**Pull request by dunglas**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | n/a
| License       | MIT
| Doc PR        | n/a

This PR enhances the whole Developper eXperience of API Platform by integrating Swagger UI more deeply:

capture d ecran 2016-09-05 a 17 31 19

* Any request having an `Accept` header with HTML set as the preferred format (usually a browser) or to an URL ending with `.html` now displays Swagger UI
* The result of the request in the default format of the API (usually JSON-LD) is shown in Swagger UI
* The doc of the requested endpoint is opened (docs for all other endpoints are collapsed) and the browser scrolls automatically to it

Alternate formats available for the current document are also directly listed from this page and can be downloaded from the browser:
capture d ecran 2016-09-05 a 17 33 02

Last but not least, if the request format is set to HTML (usually, a browser) and an exception occurs, the default Symfony's exception page is displayed. In all other cases, our JSON error mechanism is used to display the error serialized in the API Problem format or the Hydra format depending of the requested MIME type.
capture d ecran 2016-09-05 a 17 34 23

### CODE REVIEW

File: src/EventListener/ExceptionListener.php [hunk 1]
```diff
@@ -26,7 +26,10 @@ public function onKernelException(GetResponseForExceptionEvent $event)
     {
         $request = $event->getRequest();
         // Normalize exceptions only for routes managed by API Platform
-        if (!$request->attributes->has('_api_resource_class') && !$request->attributes->has('_api_respond')) {
+        if (
+            (!$request->attributes->has('_api_resource_class') && !$request->attributes->has('_api_respond')) ||
```
Comment (Simperfit) [inline src/EventListener/ExceptionListener.php]: `null ===` ?

File: src/Bridge/Symfony/Bundle/Resources/views/SwaggerUi/index.html.twig [hunk 2]
```diff
@@ -28,22 +28,37 @@
     <script src="{{ asset('bundles/apiplatform/swagger-ui/lib/swagger-oauth.js') }}"></script>
 
     <script>
+        {% if operationId %}
```
Comment (Simperfit) [inline src/Bridge/Symfony/Bundle/Resources/views/SwaggerUi/index.html.twig]: what is the purpose of that ?

Comment (dunglas) [inline src/Bridge/Symfony/Bundle/Resources/views/SwaggerUi/index.html.twig]: `operationId` may be `null` (`/doc`, `/index`).

Comment (dunglas) [inline src/Bridge/Symfony/Bundle/Resources/views/SwaggerUi/index.html.twig]: Ok got it, need to be removed :)

File: src/EventListener/ExceptionListener.php (hunk 1 above)
Comment (dunglas) [inline src/EventListener/ExceptionListener.php]: The `has` method returns a boolean.

File: src/Bridge/Symfony/Bundle/Resources/views/SwaggerUi/index.html.twig (hunk 2 above)
Comment (Simperfit) [inline src/Bridge/Symfony/Bundle/Resources/views/SwaggerUi/index.html.twig]: This is what I thought ;)

File: src/EventListener/ExceptionListener.php (hunk 1 above)
Comment (Simperfit) [inline src/EventListener/ExceptionListener.php]: Oh, I didn't see that you used has here, why are you using get up here ?

Comment (dunglas) [inline src/EventListener/ExceptionListener.php]: Good catch, `has` should be used in the other class.

**Comment by teohhanhui** [comment]
I have really mixed feelings about this one.

**Comment by dunglas** [comment]
@teohhanhui why? I was just thinking about adding a config flag for this feature.

**Comment by teohhanhui** [comment]
Nobody really expects HTML from an API. Yes, the browser asking for it, but
who actually wants that from an API? For example, it makes browser
extensions for JSON viewing useless.

But more importantly, it's wrong on principle. Showing the documentation
instead of a resource (or collection) - does that really count as just a
different representation? I think it has crossed the line.

**Comment by dunglas** [comment]
It's more a developper tool (and especially a client developper tool) than a real representation of the resource. As you pointed out, it allows to interact with the API even if you don't have Postman or a similar tool installed. It also allows to discover the API (and API Platform) with only a browser (and even a smartphone).

Django REST as a somewhat similar feature and it's pleasant to use: http://restframework.herokuapp.com/

But I got your point, it's borderline regarding the REST philosophy (some may argue that this is a real HTML representation of the resource, the data is technically displayed as HTML). WDYT about turning this feature on by default but having a config option to easily disable it?

**Comment by soyuka** [comment]
> WDYT about keeping this feature on by default but having a config option to easily disable it?

It's necessary to be able to disable it, from my point of view ofc. I'd have it disabled by default but I also understand that it's an attractive feature which is great to have with zero configuration.

**Comment by teohhanhui** [comment]
I can't pass judgement on the DX until I've actually used it (probably soon).

**Comment by teohhanhui** [comment]
Just tried it. The DX leaves much to be desired. There is an additional delay which is really annoying, also Swagger UI is not very helpful for discovery, but most importantly it prevents tools from working.

**Comment by teohhanhui** [comment]
But on the other hand, disabling it is already possible by setting the `api_platform.formats` config to not include html.

**Pull request merged**